<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>knowseqpy.calculate_gene_expression_values &mdash; KnowSeqPy  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            KnowSeqPy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <!-- Local TOC -->
              <div class="local-toc"></div>
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">KnowSeqPy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">knowseqpy.calculate_gene_expression_values</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for knowseqpy.calculate_gene_expression_values</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This module offers functionality to calculate gene expression values from RNA-seq data.</span>
<span class="sd">It supports both human and non-human gene lengths. It processes a counts and annotation DataFrames</span>
<span class="sd">to produce a DataFrame of gene expression values.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">tempfile</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">from</span> <span class="nn">.utils</span> <span class="kn">import</span> <span class="n">dataframe_to_feather</span><span class="p">,</span> <span class="n">feather_to_dataframe</span><span class="p">,</span> <span class="n">get_logger</span><span class="p">,</span> <span class="n">get_project_path</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">get_logger</span><span class="p">()</span><span class="o">.</span><span class="n">getChild</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>


<div class="viewcode-block" id="calculate_gene_expression_values">
<a class="viewcode-back" href="../../knowseqpy.html#knowseqpy.calculate_gene_expression_values.calculate_gene_expression_values">[docs]</a>
<span class="k">def</span> <span class="nf">calculate_gene_expression_values</span><span class="p">(</span><span class="n">counts</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">gene_annotation</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">genes_names</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
                                     <span class="n">not_human_gene_length_csv</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">ensembl_id</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculates the gene expression values by using a matrix of counts from RNA-seq.</span>

<span class="sd">    Args:</span>
<span class="sd">        counts: The counts pd.DataFrame with genes in rows and samples in columns.</span>
<span class="sd">        gene_annotation: A DataFrame containing gene annotations.</span>
<span class="sd">        genes_names: Use gene names instead of Ensembl IDs.</span>
<span class="sd">        not_human_gene_length_csv: Path to the gene length file if the data is not from humans.</span>
<span class="sd">        ensembl_id: Whether the counts pd.DataFrame contains Ensembl IDs.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: A pd.DataFrame containing the gene expression values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">ensembl_id</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;Non ensembl_id gene extraction has not been implemented yet&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">not_human_gene_length_csv</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
        <span class="c1"># Load default human gene length data</span>
        <span class="n">genes_length_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="vm">__file__</span><span class="p">)</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span><span class="o">.</span><span class="n">parent</span> <span class="o">/</span> <span class="s2">&quot;external_data&quot;</span> <span class="o">/</span> <span class="s2">&quot;genes_length_homo_sapiens.csv&quot;</span>
        <span class="n">gene_length</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">genes_length_path</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="s2">&quot;Gene_stable_ID&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">Path</span><span class="p">(</span><span class="n">not_human_gene_length_csv</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s2">&quot;Not Human gene length CSV file not found, please revise the path to the file.&quot;</span><span class="p">)</span>

        <span class="n">gene_length</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">not_human_gene_length_csv</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Calculating gene expression values...&quot;</span><span class="p">)</span>

    <span class="c1"># Remove duplicates from gene_annotation, as we&#39;ll only need (ensembl_gene_id, percentage_gene_gc_content) pairs</span>
    <span class="n">genes_annot</span> <span class="o">=</span> <span class="n">gene_annotation</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">gene_annotation</span><span class="o">.</span><span class="n">index</span><span class="p">)</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>

    <span class="c1"># We get the gene_annotations and length only for those genes that are present in counts_df</span>
    <span class="n">common_joined</span> <span class="o">=</span> <span class="n">genes_annot</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_length</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">)</span>
    <span class="n">common_genes_annot</span> <span class="o">=</span> <span class="n">common_joined</span><span class="p">[</span><span class="s2">&quot;percentage_gene_gc_content&quot;</span><span class="p">]</span>
    <span class="n">common_gene_length</span> <span class="o">=</span> <span class="n">common_joined</span><span class="p">[</span><span class="s2">&quot;Gene_length&quot;</span><span class="p">]</span>
    <span class="n">common_joined</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">genes_annot</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">gene_length</span><span class="o">.</span><span class="n">columns</span><span class="p">),</span>
                       <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Calculating gene expression values using R&quot;</span><span class="p">)</span>
    <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">TemporaryDirectory</span><span class="p">()</span> <span class="k">as</span> <span class="n">temp_dir</span><span class="p">:</span>
        <span class="n">counts_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">temp_dir</span><span class="p">,</span> <span class="s2">&quot;counts.feather&quot;</span><span class="p">)</span>
        <span class="n">x_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">temp_dir</span><span class="p">,</span> <span class="s2">&quot;x.feather&quot;</span><span class="p">)</span>
        <span class="n">length_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">temp_dir</span><span class="p">,</span> <span class="s2">&quot;length.feather&quot;</span><span class="p">)</span>
        <span class="n">results_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">temp_dir</span><span class="p">,</span> <span class="s2">&quot;gene_results.feather&quot;</span><span class="p">)</span>

        <span class="n">dataframe_to_feather</span><span class="p">(</span><span class="n">common_joined</span><span class="p">,</span> <span class="n">counts_path</span><span class="p">)</span>
        <span class="n">dataframe_to_feather</span><span class="p">(</span><span class="n">common_genes_annot</span><span class="o">.</span><span class="n">to_frame</span><span class="p">(),</span> <span class="n">x_path</span><span class="p">)</span>
        <span class="n">dataframe_to_feather</span><span class="p">(</span><span class="n">common_gene_length</span><span class="o">.</span><span class="n">to_frame</span><span class="p">(),</span> <span class="n">length_path</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span>
                <span class="s2">&quot;Rscript&quot;</span><span class="p">,</span>
                <span class="n">get_project_path</span><span class="p">()</span> <span class="o">/</span> <span class="s2">&quot;knowseqpy&quot;</span> <span class="o">/</span> <span class="s2">&quot;r_scripts&quot;</span> <span class="o">/</span> <span class="s2">&quot;cqnWorkflow.R&quot;</span><span class="p">,</span>
                <span class="nb">str</span><span class="p">(</span><span class="n">counts_path</span><span class="p">),</span>
                <span class="nb">str</span><span class="p">(</span><span class="n">x_path</span><span class="p">),</span>
                <span class="nb">str</span><span class="p">(</span><span class="n">length_path</span><span class="p">),</span>
                <span class="nb">str</span><span class="p">(</span><span class="n">results_path</span><span class="p">)</span>
            <span class="p">],</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">except</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">CalledProcessError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to execute cqn R script. </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">e</span>

        <span class="n">gene_expression</span> <span class="o">=</span> <span class="n">feather_to_dataframe</span><span class="p">(</span><span class="n">results_path</span><span class="p">)</span>
        <span class="n">gene_expression</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;row_name&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">gene_expression</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># Use gene_names. If gene ID not found in gene_annotation, its row is removed from the final df</span>
        <span class="k">if</span> <span class="n">genes_names</span><span class="p">:</span>
            <span class="n">gene_expression</span> <span class="o">=</span> <span class="n">gene_expression</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">genes_annot</span><span class="p">[</span><span class="s2">&quot;external_gene_name&quot;</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">)</span>
            <span class="n">gene_expression</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;external_gene_name&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">gene_expression</span> <span class="o">=</span> <span class="n">gene_expression</span><span class="p">[</span><span class="o">~</span><span class="n">gene_expression</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">duplicated</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">)]</span>

        <span class="k">return</span> <span class="n">gene_expression</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Blanca Domene.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>