<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>knowseqpy.degs_extraction &mdash; KnowSeqPy  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            KnowSeqPy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <!-- Local TOC -->
              <div class="local-toc"></div>
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">KnowSeqPy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">knowseqpy.degs_extraction</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for knowseqpy.degs_extraction</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This module is designed to facilitate the extraction of differentially expressed genes (DEGs) from</span>
<span class="sd">gene expression data. It leverages both Python and R to perform comprehensive DEGs analysis, allowing</span>
<span class="sd">users to identify significant changes in gene expression across different conditions or classes. The</span>
<span class="sd">module supports both biclass and (planned) multiclass analysis, integrating with the R `limma` package</span>
<span class="sd">for statistical analysis.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">tempfile</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">patsy.highlevel</span> <span class="kn">import</span> <span class="n">dmatrix</span>
<span class="kn">from</span> <span class="nn">sklearn.model_selection</span> <span class="kn">import</span> <span class="n">KFold</span>

<span class="kn">from</span> <span class="nn">.utils</span> <span class="kn">import</span> <span class="n">get_logger</span><span class="p">,</span> <span class="n">get_project_path</span><span class="p">,</span> <span class="n">dataframe_to_feather</span><span class="p">,</span> <span class="n">feather_to_dataframe</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">get_logger</span><span class="p">()</span><span class="o">.</span><span class="n">getChild</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>


<div class="viewcode-block" id="degs_extraction">
<a class="viewcode-back" href="../../knowseqpy.html#knowseqpy.degs_extraction.degs_extraction">[docs]</a>
<span class="k">def</span> <span class="nf">degs_extraction</span><span class="p">(</span><span class="n">data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">labels</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span> <span class="n">max_genes</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;inf&quot;</span><span class="p">),</span> <span class="n">p_value</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
                    <span class="n">lfc</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span> <span class="n">cv</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">k_folds</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Performs the analysis to extract Differentially Expressed Genes (DEGs) among classes to compare.</span>

<span class="sd">    Args:</span>
<span class="sd">        data: DataFrame containing genes in rows and samples in columns.</span>
<span class="sd">        labels: Series containing labels for each sample in data.</span>
<span class="sd">        p_value: p-value threshold for determining DEGs. Defaults to 0.05.</span>
<span class="sd">        lfc: Log Fold Change threshold for determining DEGs. Defaults to 1.0.</span>
<span class="sd">        max_genes: Maximum number of genes as output. Defaults to all genes.</span>
<span class="sd">        cv: If True, runs Cross-Validation DEGs extraction. Defaults to False.</span>
<span class="sd">        k_folds: Number of folds for Cross-Validation. Defaults to 5.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dictionary containing DEGs analysis results.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">labels</span> <span class="o">=</span> <span class="n">labels</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;category&quot;</span><span class="p">)</span>

    <span class="n">cv_datasets</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">cv</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Applying DEGs extraction with Cross-Validation&quot;</span><span class="p">)</span>
        <span class="n">kf</span> <span class="o">=</span> <span class="n">KFold</span><span class="p">(</span><span class="n">n_splits</span><span class="o">=</span><span class="n">k_folds</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">train_index</span><span class="p">,</span> <span class="n">test_index</span> <span class="ow">in</span> <span class="n">kf</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
            <span class="n">fold_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">train_index</span><span class="p">]</span>
            <span class="n">fold_labels</span> <span class="o">=</span> <span class="n">labels</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">test_index</span><span class="p">]</span>
            <span class="n">cv_datasets</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">fold_data</span><span class="p">,</span> <span class="n">fold_labels</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">cv_datasets</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">data</span><span class="p">,</span> <span class="n">labels</span><span class="p">))</span>

    <span class="c1"># Perform DEGs analysis for each dataset</span>
    <span class="n">cv_degs_results</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">cv_data</span><span class="p">,</span> <span class="n">cv_labels</span> <span class="ow">in</span> <span class="n">cv_datasets</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">labels</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Two classes detected, applying biclass analysis&quot;</span><span class="p">)</span>
            <span class="n">deg_res</span> <span class="o">=</span> <span class="n">_biclass_analysis</span><span class="p">(</span><span class="n">cv_data</span><span class="p">,</span> <span class="n">cv_labels</span><span class="p">,</span> <span class="n">p_value</span><span class="p">,</span> <span class="n">lfc</span><span class="p">,</span> <span class="n">max_genes</span><span class="p">)</span>
            <span class="n">cv_degs_results</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">deg_res</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">labels</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;More than two classes detected, applying multiclass analysis&quot;</span><span class="p">)</span>
            <span class="n">cv_degs_results</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_multiclass_analysis</span><span class="p">())</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Number of classes in labels must be at least 2.&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">cv_degs_results</span></div>



<span class="k">def</span> <span class="nf">_biclass_analysis</span><span class="p">(</span><span class="n">data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">labels</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span> <span class="n">p_value</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">lfc</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
                      <span class="n">max_genes</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Performs biclass DEGs analysis by calling an R script for lmFit and eBayes.</span>

<span class="sd">    Args:</span>
<span class="sd">        data: DataFrame containing genes in rows and samples in columns.</span>
<span class="sd">        labels: Series containing labels for each sample in data.</span>
<span class="sd">        p_value: p-value threshold for determining DEGs.</span>
<span class="sd">        lfc: Log Fold Change threshold for determining DEGs.</span>
<span class="sd">        max_genes: Maximum number of genes as output.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Analysis results including DEGs table and matrix.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">degs_table</span> <span class="o">=</span> <span class="n">run_limma_deg_analysis</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">labels</span><span class="p">,</span> <span class="n">p_value</span><span class="p">,</span> <span class="n">lfc</span><span class="p">,</span> <span class="n">max_genes</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">max_genes</span> <span class="o">!=</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;inf&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">degs_table</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">max_genes</span><span class="p">:</span>
        <span class="n">degs_table</span> <span class="o">=</span> <span class="n">degs_table</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">max_genes</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">degs_table</span>


<div class="viewcode-block" id="run_limma_deg_analysis">
<a class="viewcode-back" href="../../knowseqpy.html#knowseqpy.degs_extraction.run_limma_deg_analysis">[docs]</a>
<span class="k">def</span> <span class="nf">run_limma_deg_analysis</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">labels</span><span class="p">,</span> <span class="n">p_value</span><span class="p">,</span> <span class="n">lfc</span><span class="p">,</span> <span class="n">max_genes</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Runs differential expression analysis using R&#39;s limma package.</span>

<span class="sd">    Args:</span>
<span class="sd">        data: A pandas DataFrame containing gene expression data.</span>
<span class="sd">        labels: Series containing labels for each sample in data.</span>
<span class="sd">        p_value: p-value threshold for determining DEGs.</span>
<span class="sd">        lfc: Log Fold Change threshold for determining DEGs.</span>
<span class="sd">        max_genes: Maximum number of genes as output.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A pandas DataFrame containing the limma top_table results.</span>

<span class="sd">    Raises:</span>
<span class="sd">        RuntimeError: If external R script execution fails.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">labels</span> <span class="o">=</span> <span class="n">labels</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;category&quot;</span><span class="p">)</span>

    <span class="n">transposed_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span><span class="o">.</span><span class="n">T</span>
    <span class="n">transposed_data</span><span class="p">[</span><span class="s2">&quot;sample_class&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">labels</span><span class="o">.</span><span class="n">values</span>
    <span class="n">design_matrix</span> <span class="o">=</span> <span class="n">dmatrix</span><span class="p">(</span><span class="n">formula_like</span><span class="o">=</span><span class="s2">&quot;1 + C(sample_class)&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">transposed_data</span><span class="p">,</span> <span class="n">return_type</span><span class="o">=</span><span class="s2">&quot;dataframe&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span>
        <span class="nb">int</span><span class="p">)</span>

    <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">TemporaryDirectory</span><span class="p">()</span> <span class="k">as</span> <span class="n">temp_dir</span><span class="p">:</span>
        <span class="n">expression_data_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">temp_dir</span><span class="p">,</span> <span class="s2">&quot;expression_data.feather&quot;</span><span class="p">)</span>
        <span class="n">design_matrix_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">temp_dir</span><span class="p">,</span> <span class="s2">&quot;design_matrix.feather&quot;</span><span class="p">)</span>
        <span class="n">limma_results_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">temp_dir</span><span class="p">,</span> <span class="s2">&quot;limma_results.feather&quot;</span><span class="p">)</span>

        <span class="n">dataframe_to_feather</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">expression_data_path</span><span class="p">)</span>
        <span class="n">dataframe_to_feather</span><span class="p">(</span><span class="n">design_matrix</span><span class="p">,</span> <span class="n">design_matrix_path</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span>
                <span class="s2">&quot;Rscript&quot;</span><span class="p">,</span>
                <span class="n">get_project_path</span><span class="p">()</span> <span class="o">/</span> <span class="s2">&quot;knowseqpy&quot;</span> <span class="o">/</span> <span class="s2">&quot;r_scripts&quot;</span> <span class="o">/</span> <span class="s2">&quot;LimmaDEGsExtractionWorkflow.R&quot;</span><span class="p">,</span>
                <span class="n">expression_data_path</span><span class="p">,</span>
                <span class="n">design_matrix_path</span><span class="p">,</span>
                <span class="n">limma_results_path</span><span class="p">,</span>
                <span class="nb">str</span><span class="p">(</span><span class="n">p_value</span><span class="p">),</span>
                <span class="nb">str</span><span class="p">(</span><span class="n">lfc</span><span class="p">),</span>
                <span class="nb">str</span><span class="p">(</span><span class="n">max_genes</span><span class="p">)</span>
            <span class="p">],</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">except</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">CalledProcessError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to execute DEGs extraction R script. </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">e</span>

        <span class="n">results</span> <span class="o">=</span> <span class="n">feather_to_dataframe</span><span class="p">(</span><span class="n">limma_results_path</span><span class="p">)</span>
        <span class="n">results</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;row_name&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">results</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">return</span> <span class="n">results</span></div>



<span class="k">def</span> <span class="nf">_multiclass_analysis</span><span class="p">():</span>
    <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;Multiclass analysis function has not been implemented yet&quot;</span><span class="p">)</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Blanca Domene.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>