<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>knowseqpy package &mdash; KnowSeqPy  documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=5929fcd5"></script>
        <script src="_static/doctools.js?v=9a2dae69"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            KnowSeqPy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">knowseqpy package</a><ul>
<li><a class="reference internal" href="#subpackages">Subpackages</a></li>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-knowseqpy.calculate_gene_expression_values">knowseqpy.calculate_gene_expression_values module</a><ul>
<li><a class="reference internal" href="#knowseqpy.calculate_gene_expression_values.calculate_gene_expression_values"><code class="docutils literal notranslate"><span class="pre">calculate_gene_expression_values()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-knowseqpy.counts_to_dataframe">knowseqpy.counts_to_dataframe module</a><ul>
<li><a class="reference internal" href="#knowseqpy.counts_to_dataframe.counts_to_dataframe"><code class="docutils literal notranslate"><span class="pre">counts_to_dataframe()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-knowseqpy.degs_extraction">knowseqpy.degs_extraction module</a><ul>
<li><a class="reference internal" href="#knowseqpy.degs_extraction.degs_extraction"><code class="docutils literal notranslate"><span class="pre">degs_extraction()</span></code></a></li>
<li><a class="reference internal" href="#knowseqpy.degs_extraction.run_limma_deg_analysis"><code class="docutils literal notranslate"><span class="pre">run_limma_deg_analysis()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-knowseqpy.evaluate_model">knowseqpy.evaluate_model module</a><ul>
<li><a class="reference internal" href="#knowseqpy.evaluate_model.evaluate_model"><code class="docutils literal notranslate"><span class="pre">evaluate_model()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-knowseqpy.get_genes_annotation">knowseqpy.get_genes_annotation module</a><ul>
<li><a class="reference internal" href="#knowseqpy.get_genes_annotation.get_genes_annotation"><code class="docutils literal notranslate"><span class="pre">get_genes_annotation()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-knowseqpy.read_dge">knowseqpy.read_dge module</a><ul>
<li><a class="reference internal" href="#knowseqpy.read_dge.read_dge"><code class="docutils literal notranslate"><span class="pre">read_dge()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-knowseqpy.rna_seq_qa">knowseqpy.rna_seq_qa module</a><ul>
<li><a class="reference internal" href="#knowseqpy.rna_seq_qa.rna_seq_qa"><code class="docutils literal notranslate"><span class="pre">rna_seq_qa()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-knowseqpy">Module contents</a></li>
</ul>
</li>
</ul>
</div>
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">KnowSeqPy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">knowseqpy package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/knowseqpy.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="knowseqpy-package">
<h1>knowseqpy package<a class="headerlink" href="#knowseqpy-package" title="Link to this heading"></a></h1>
<section id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Link to this heading"></a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="knowseqpy.batch_effect.html">knowseqpy.batch_effect package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.batch_effect.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.batch_effect.html#module-knowseqpy.batch_effect.combat">knowseqpy.batch_effect.combat module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.batch_effect.html#knowseqpy.batch_effect.combat.combat"><code class="docutils literal notranslate"><span class="pre">combat()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.batch_effect.html#module-knowseqpy.batch_effect.sva">knowseqpy.batch_effect.sva module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.batch_effect.html#knowseqpy.batch_effect.sva.sva"><code class="docutils literal notranslate"><span class="pre">sva()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.batch_effect.html#module-knowseqpy.batch_effect">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="knowseqpy.classifiers.html">knowseqpy.classifiers package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.classifiers.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.classifiers.html#module-knowseqpy.classifiers.decision_tree">knowseqpy.classifiers.decision_tree module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.classifiers.html#knowseqpy.classifiers.decision_tree.decision_tree"><code class="docutils literal notranslate"><span class="pre">decision_tree()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.classifiers.html#module-knowseqpy.classifiers.gradient_boosting">knowseqpy.classifiers.gradient_boosting module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.classifiers.html#knowseqpy.classifiers.gradient_boosting.gradient_boosting"><code class="docutils literal notranslate"><span class="pre">gradient_boosting()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.classifiers.html#module-knowseqpy.classifiers.keras_nn">knowseqpy.classifiers.keras_nn module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.classifiers.html#knowseqpy.classifiers.keras_nn.create_nn_model"><code class="docutils literal notranslate"><span class="pre">create_nn_model()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.classifiers.html#knowseqpy.classifiers.keras_nn.nn"><code class="docutils literal notranslate"><span class="pre">nn()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.classifiers.html#module-knowseqpy.classifiers.knn">knowseqpy.classifiers.knn module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.classifiers.html#knowseqpy.classifiers.knn.knn"><code class="docutils literal notranslate"><span class="pre">knn()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.classifiers.html#module-knowseqpy.classifiers.logistic_regression">knowseqpy.classifiers.logistic_regression module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.classifiers.html#knowseqpy.classifiers.logistic_regression.logistic_regression"><code class="docutils literal notranslate"><span class="pre">logistic_regression()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.classifiers.html#module-knowseqpy.classifiers.rf">knowseqpy.classifiers.rf module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.classifiers.html#knowseqpy.classifiers.rf.rf"><code class="docutils literal notranslate"><span class="pre">rf()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.classifiers.html#module-knowseqpy.classifiers.svm">knowseqpy.classifiers.svm module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.classifiers.html#knowseqpy.classifiers.svm.svm"><code class="docutils literal notranslate"><span class="pre">svm()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.classifiers.html#module-knowseqpy.classifiers">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="knowseqpy.feature_selection.html">knowseqpy.feature_selection package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.feature_selection.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.feature_selection.html#module-knowseqpy.feature_selection.linear_discriminant_analysis">knowseqpy.feature_selection.linear_discriminant_analysis module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.feature_selection.html#knowseqpy.feature_selection.linear_discriminant_analysis.linear_discriminant_analysis"><code class="docutils literal notranslate"><span class="pre">linear_discriminant_analysis()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.feature_selection.html#module-knowseqpy.feature_selection.mrmr">knowseqpy.feature_selection.mrmr module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.feature_selection.html#knowseqpy.feature_selection.mrmr.mrmr"><code class="docutils literal notranslate"><span class="pre">mrmr()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.feature_selection.html#module-knowseqpy.feature_selection.random_forest">knowseqpy.feature_selection.random_forest module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.feature_selection.html#knowseqpy.feature_selection.random_forest.random_forest"><code class="docutils literal notranslate"><span class="pre">random_forest()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.feature_selection.html#module-knowseqpy.feature_selection">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="knowseqpy.normalization.html">knowseqpy.normalization package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.normalization.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.normalization.html#module-knowseqpy.normalization.cpm">knowseqpy.normalization.cpm module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.normalization.html#knowseqpy.normalization.cpm.cpm"><code class="docutils literal notranslate"><span class="pre">cpm()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.normalization.html#module-knowseqpy.normalization.cqn">knowseqpy.normalization.cqn module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.normalization.html#knowseqpy.normalization.cqn.cqn"><code class="docutils literal notranslate"><span class="pre">cqn()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.normalization.html#module-knowseqpy.normalization">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="knowseqpy.outlier_detection.html">knowseqpy.outlier_detection package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.outlier_detection.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.outlier_detection.html#module-knowseqpy.outlier_detection.kolmogorov_smirnov">knowseqpy.outlier_detection.kolmogorov_smirnov module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.outlier_detection.html#knowseqpy.outlier_detection.kolmogorov_smirnov.kolmogorov_smirnov"><code class="docutils literal notranslate"><span class="pre">kolmogorov_smirnov()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.outlier_detection.html#module-knowseqpy.outlier_detection.manhattan_distance">knowseqpy.outlier_detection.manhattan_distance module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.outlier_detection.html#knowseqpy.outlier_detection.manhattan_distance.manhattan_distance"><code class="docutils literal notranslate"><span class="pre">manhattan_distance()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.outlier_detection.html#module-knowseqpy.outlier_detection.median_deviation">knowseqpy.outlier_detection.median_deviation module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.outlier_detection.html#knowseqpy.outlier_detection.median_deviation.median_deviation"><code class="docutils literal notranslate"><span class="pre">median_deviation()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.outlier_detection.html#module-knowseqpy.outlier_detection">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="knowseqpy.utils.html">knowseqpy.utils package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.utils.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.utils.html#module-knowseqpy.utils.common">knowseqpy.utils.common module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.utils.html#knowseqpy.utils.common.calculate_specificity"><code class="docutils literal notranslate"><span class="pre">calculate_specificity()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.utils.html#knowseqpy.utils.common.csv_to_dataframe"><code class="docutils literal notranslate"><span class="pre">csv_to_dataframe()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.utils.html#knowseqpy.utils.common.csv_to_list"><code class="docutils literal notranslate"><span class="pre">csv_to_list()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.utils.html#knowseqpy.utils.common.dataframe_to_feather"><code class="docutils literal notranslate"><span class="pre">dataframe_to_feather()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.utils.html#knowseqpy.utils.common.feather_to_dataframe"><code class="docutils literal notranslate"><span class="pre">feather_to_dataframe()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.utils.html#knowseqpy.utils.common.get_project_path"><code class="docutils literal notranslate"><span class="pre">get_project_path()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.utils.html#knowseqpy.utils.common.get_test_path"><code class="docutils literal notranslate"><span class="pre">get_test_path()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.utils.html#module-knowseqpy.utils.logger">knowseqpy.utils.logger module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.utils.html#knowseqpy.utils.logger.get_logger"><code class="docutils literal notranslate"><span class="pre">get_logger()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.utils.html#knowseqpy.utils.logger.set_logger"><code class="docutils literal notranslate"><span class="pre">set_logger()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.utils.html#module-knowseqpy.utils.plotting">knowseqpy.utils.plotting module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.utils.html#knowseqpy.utils.plotting.another_plot_decision_tree"><code class="docutils literal notranslate"><span class="pre">another_plot_decision_tree()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.utils.html#knowseqpy.utils.plotting.plot_boxplot"><code class="docutils literal notranslate"><span class="pre">plot_boxplot()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.utils.html#knowseqpy.utils.plotting.plot_confusion_matrix"><code class="docutils literal notranslate"><span class="pre">plot_confusion_matrix()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.utils.html#knowseqpy.utils.plotting.plot_decision_boundary"><code class="docutils literal notranslate"><span class="pre">plot_decision_boundary()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.utils.html#knowseqpy.utils.plotting.plot_decision_tree"><code class="docutils literal notranslate"><span class="pre">plot_decision_tree()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="knowseqpy.utils.html#knowseqpy.utils.plotting.plot_samples_heatmap"><code class="docutils literal notranslate"><span class="pre">plot_samples_heatmap()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="knowseqpy.utils.html#module-knowseqpy.utils">Module contents</a></li>
</ul>
</li>
</ul>
</div>
</section>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading"></a></h2>
</section>
<section id="module-knowseqpy.calculate_gene_expression_values">
<span id="knowseqpy-calculate-gene-expression-values-module"></span><h2>knowseqpy.calculate_gene_expression_values module<a class="headerlink" href="#module-knowseqpy.calculate_gene_expression_values" title="Link to this heading"></a></h2>
<p>This module offers functionality to calculate gene expression values from RNA-seq data.
It supports both human and non-human gene lengths. It processes a counts and annotation DataFrames
to produce a DataFrame of gene expression values.</p>
<dl class="py function">
<dt class="sig sig-object py" id="knowseqpy.calculate_gene_expression_values.calculate_gene_expression_values">
<span class="sig-prename descclassname"><span class="pre">knowseqpy.calculate_gene_expression_values.</span></span><span class="sig-name descname"><span class="pre">calculate_gene_expression_values</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">counts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">not_human_gene_length_csv</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/knowseqpy/calculate_gene_expression_values.html#calculate_gene_expression_values"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#knowseqpy.calculate_gene_expression_values.calculate_gene_expression_values" title="Link to this definition"></a></dt>
<dd><p>Calculates the gene expression values by using a matrix of counts from RNA-seq.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>counts</strong> – The counts pd.DataFrame with genes in rows and samples in columns.</p></li>
<li><p><strong>gene_annotation</strong> – A DataFrame containing gene annotations.</p></li>
<li><p><strong>genes_names</strong> – Use gene names instead of Ensembl IDs.</p></li>
<li><p><strong>not_human_gene_length_csv</strong> – Path to the gene length file if the data is not from humans.</p></li>
<li><p><strong>ensembl_id</strong> – Whether the counts pd.DataFrame contains Ensembl IDs.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A pd.DataFrame containing the gene expression values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-knowseqpy.counts_to_dataframe">
<span id="knowseqpy-counts-to-dataframe-module"></span><h2>knowseqpy.counts_to_dataframe module<a class="headerlink" href="#module-knowseqpy.counts_to_dataframe" title="Link to this heading"></a></h2>
<p>This module contains functions for processing RNA-Seq gene expression count files, compiling gene count data from
multiple files into a single DataFrame, using CPM and filtering out unwanted data based on specified criteria.</p>
<dl class="py function">
<dt class="sig sig-object py" id="knowseqpy.counts_to_dataframe.counts_to_dataframe">
<span class="sig-prename descclassname"><span class="pre">knowseqpy.counts_to_dataframe.</span></span><span class="sig-name descname"><span class="pre">counts_to_dataframe</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">info_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">counts_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">','</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ext</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'.count'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rows_to_skip</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Series</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/knowseqpy/counts_to_dataframe.html#counts_to_dataframe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#knowseqpy.counts_to_dataframe.counts_to_dataframe" title="Link to this definition"></a></dt>
<dd><p>Merges the information of all count files found in <cite>counts_path</cite>, which ID correspond to the one found
in <cite>info_path[“Internal.ID”]</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>info_path</strong> – path to CSV file that holds metadata for biological samples. Each row represents a unique sample
with various identifiers and descriptors. The expected columns are <cite>Internal.ID</cite> and <cite>Sample.Type</cite>.</p></li>
<li><p><strong>counts_path</strong> – The directory path where <cite>.count</cite> files are stored. Each file contains gene expression data
in count format for a sample identified by Internal.ID.</p></li>
<li><p><strong>sep</strong> – The separator character of the <cite>info_path</cite> content.</p></li>
<li><p><strong>ext</strong> – The extension of the count file.</p></li>
<li><p><strong>rows_to_skip</strong> – List of rows to skip during processing. Defaults to common RNA-Seq metadata rows.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Pandas dataframe with gene ids (ensemble ID) as rows and sample ids of each count file as columns.
Pandas series with the classes that correspond to each sample.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-knowseqpy.degs_extraction">
<span id="knowseqpy-degs-extraction-module"></span><h2>knowseqpy.degs_extraction module<a class="headerlink" href="#module-knowseqpy.degs_extraction" title="Link to this heading"></a></h2>
<p>This module is designed to facilitate the extraction of differentially expressed genes (DEGs) from
gene expression data. It leverages both Python and R to perform comprehensive DEGs analysis, allowing
users to identify significant changes in gene expression across different conditions or classes. The
module supports both biclass and (planned) multiclass analysis, integrating with the R <cite>limma</cite> package
for statistical analysis.</p>
<dl class="py function">
<dt class="sig sig-object py" id="knowseqpy.degs_extraction.degs_extraction">
<span class="sig-prename descclassname"><span class="pre">knowseqpy.degs_extraction.</span></span><span class="sig-name descname"><span class="pre">degs_extraction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Series</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">inf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p_value</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lfc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cv</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k_folds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/knowseqpy/degs_extraction.html#degs_extraction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#knowseqpy.degs_extraction.degs_extraction" title="Link to this definition"></a></dt>
<dd><p>Performs the analysis to extract Differentially Expressed Genes (DEGs) among classes to compare.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – DataFrame containing genes in rows and samples in columns.</p></li>
<li><p><strong>labels</strong> – Series containing labels for each sample in data.</p></li>
<li><p><strong>p_value</strong> – p-value threshold for determining DEGs. Defaults to 0.05.</p></li>
<li><p><strong>lfc</strong> – Log Fold Change threshold for determining DEGs. Defaults to 1.0.</p></li>
<li><p><strong>max_genes</strong> – Maximum number of genes as output. Defaults to all genes.</p></li>
<li><p><strong>cv</strong> – If True, runs Cross-Validation DEGs extraction. Defaults to False.</p></li>
<li><p><strong>k_folds</strong> – Number of folds for Cross-Validation. Defaults to 5.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary containing DEGs analysis results.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="knowseqpy.degs_extraction.run_limma_deg_analysis">
<span class="sig-prename descclassname"><span class="pre">knowseqpy.degs_extraction.</span></span><span class="sig-name descname"><span class="pre">run_limma_deg_analysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p_value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lfc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_genes</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/knowseqpy/degs_extraction.html#run_limma_deg_analysis"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#knowseqpy.degs_extraction.run_limma_deg_analysis" title="Link to this definition"></a></dt>
<dd><p>Runs differential expression analysis using R’s limma package.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – A pandas DataFrame containing gene expression data.</p></li>
<li><p><strong>labels</strong> – Series containing labels for each sample in data.</p></li>
<li><p><strong>p_value</strong> – p-value threshold for determining DEGs.</p></li>
<li><p><strong>lfc</strong> – Log Fold Change threshold for determining DEGs.</p></li>
<li><p><strong>max_genes</strong> – Maximum number of genes as output.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A pandas DataFrame containing the limma top_table results.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>RuntimeError</strong> – If external R script execution fails.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-knowseqpy.evaluate_model">
<span id="knowseqpy-evaluate-model-module"></span><h2>knowseqpy.evaluate_model module<a class="headerlink" href="#module-knowseqpy.evaluate_model" title="Link to this heading"></a></h2>
<p>This module provides a function to evaluate models on test data.</p>
<dl class="py function">
<dt class="sig sig-object py" id="knowseqpy.evaluate_model.evaluate_model">
<span class="sig-prename descclassname"><span class="pre">knowseqpy.evaluate_model.</span></span><span class="sig-name descname"><span class="pre">evaluate_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Pipeline</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_test</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels_test</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Series</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vars_selected</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/knowseqpy/evaluate_model.html#evaluate_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#knowseqpy.evaluate_model.evaluate_model" title="Link to this definition"></a></dt>
<dd><p>Evaluates a trained model pipeline on the test data and returns performance metrics.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – A trained model pipeline that includes preprocessing and the classifier.</p></li>
<li><p><strong>data_test</strong> – Test data (DataFrame) on which to evaluate the model.</p></li>
<li><p><strong>labels_test</strong> – True labels for the test data.</p></li>
<li><p><strong>vars_selected</strong> – Selected genes for classification. Can be DEGs or a custom list.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>accuracy, precision, recall, F1 score, and confusion matrix.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>A dictionary containing key performance metrics</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-knowseqpy.get_genes_annotation">
<span id="knowseqpy-get-genes-annotation-module"></span><h2>knowseqpy.get_genes_annotation module<a class="headerlink" href="#module-knowseqpy.get_genes_annotation" title="Link to this heading"></a></h2>
<p>This module provides functionality to retrieve gene annotations from the Ensembl biomart database. It supports querying
for specific genes or the entire genome, with customizable attributes and filters. Annotations can be fetched for both
human and non-human species, accommodating different reference genomes (GRCh37, GRCh38). Results are returned as pandas
DataFrames, facilitating further data analysis and manipulation.</p>
<dl class="py function">
<dt class="sig sig-object py" id="knowseqpy.get_genes_annotation.get_genes_annotation">
<span class="sig-prename descclassname"><span class="pre">knowseqpy.get_genes_annotation.</span></span><span class="sig-name descname"><span class="pre">get_genes_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attributes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attribute_filter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ensembl_gene_id'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">not_hsapiens_dataset</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">38</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/knowseqpy/get_genes_annotation.html#get_genes_annotation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#knowseqpy.get_genes_annotation.get_genes_annotation" title="Link to this definition"></a></dt>
<dd><p>Retrieves gene annotations from the Ensembl biomart database.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>values</strong> – A list of genes that can contain Ensembl IDs, gene names, or “allGenome”
to indicate all genes.</p></li>
<li><p><strong>attributes</strong> – A list of desired attributes or information to retrieve from Ensembl biomart.
Default is [‘ensembl_gene_id’, ‘external_gene_name’, ‘percentage_gene_gc_content’, ‘entrezgene_id’].</p></li>
<li><p><strong>attribute_filter</strong> – An attribute used as a filter to return the rest of the attributes.
Default is ‘ensembl_gene_id’.</p></li>
<li><p><strong>not_hsapiens_dataset</strong> – The dataset identification for non-human annotations. Default is None.</p></li>
<li><p><strong>reference_genome</strong> – The reference genome to use. It must be 37 or 38. Default is 38.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the requested gene annotations.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ValueError</strong> – If invalid input is provided for the parameters.</p></li>
<li><p><strong>ValueError</strong> – If an error occurs during the query, or the query result is empty or contains an error message.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-knowseqpy.read_dge">
<span id="knowseqpy-read-dge-module"></span><h2>knowseqpy.read_dge module<a class="headerlink" href="#module-knowseqpy.read_dge" title="Link to this heading"></a></h2>
<p>This module is designed to handle the processing and analysis of gene expression count data,
primarily sourced from RNA sequencing experiments. It provides functionality to read count data from
text files, merge multiple samples into a comprehensive pandas DataFrame, and perform basic preprocessing
such as normalization and identification of metadata tags.</p>
<dl class="py function">
<dt class="sig sig-object py" id="knowseqpy.read_dge.read_dge">
<span class="sig-prename descclassname"><span class="pre">knowseqpy.read_dge.</span></span><span class="sig-name descname"><span class="pre">read_dge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_info</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">counts_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ext</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'.count'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Series</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/knowseqpy/read_dge.html#read_dge"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#knowseqpy.read_dge.read_dge" title="Link to this definition"></a></dt>
<dd><p>Reads and merges a set of text files containing gene expression counts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_info</strong> – Pandas DataFrame of filenames that contain sample information.</p></li>
<li><p><strong>counts_path</strong> – The directory path where <cite>.count</cite> files are stored. Each file contains gene expression data
in count format for a sample identified by Internal.ID.</p></li>
<li><p><strong>ext</strong> – The extension of the count file.</p></li>
<li><p><strong>labels</strong> – Custom labels to associate with the files. Defaults to filenames if None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A pandas DataFrame, containing a row for each unique gene identifier found in the input files and
a column for each sample, with expression counts as values.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>Exception</strong> – If row names (gene identifiers) are not unique within any file.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-knowseqpy.rna_seq_qa">
<span id="knowseqpy-rna-seq-qa-module"></span><h2>knowseqpy.rna_seq_qa module<a class="headerlink" href="#module-knowseqpy.rna_seq_qa" title="Link to this heading"></a></h2>
<p>This module provides functionalities for performing quality analysis on RNA sequencing data.
It includes methods for detecting outliers in gene expression data using various statistical techniques,
such as the Kolmogorov-Smirnov test, Median Absolute Deviation, and Manhattan distance analysis.</p>
<dl class="py function">
<dt class="sig sig-object py" id="knowseqpy.rna_seq_qa.rna_seq_qa">
<span class="sig-prename descclassname"><span class="pre">knowseqpy.rna_seq_qa.</span></span><span class="sig-name descname"><span class="pre">rna_seq_qa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_expression_df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/knowseqpy/rna_seq_qa.html#rna_seq_qa"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#knowseqpy.rna_seq_qa.rna_seq_qa" title="Link to this definition"></a></dt>
<dd><p>Perform the quality analysis of an expression df.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>gene_expression_df</strong> – A DataFrame that contains the gene expression values.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A tuple where the first element is a DataFrame with outliers removed based on the</dt><dd><p>commonality between at least two of three methods (KS test, MAD, and Manhattan distances),
and the second element is a list containing the identifiers of the detected outliers.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-knowseqpy">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-knowseqpy" title="Link to this heading"></a></h2>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Blanca Domene.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>